tests/QueryCompatibleInputMiddlewareTest.php

```diff
@@ -0,0 +1,256 @@
+            'FooOperation',
+            [$inputParam => $inputValue]
+        );
+
+        $list = new HandlerList();
+        $list->setHandler(function ($command) {
+            return "success";
+        });
+        $list->appendValidate(QueryCompatibleInputMiddleware::wrap($service));
+
+        $handler = $list->resolve();
+        $handler($command, new Request('POST', 'https://foo.com'));
+    }
+
+    /**
+     * @dataProvider getInputs()
+     *
+     * @param $inputParam
+     * @param $inputValue
+     * @param $expected
+     */
+    public function testAppliesMiddlewareAndCastsValues($inputParam, $inputValue, $expected)
+    {
+        $service = $this->generateTestService();
+        $client = $this->generateTestClient($service);
+        $command = $client->getCommand(
+            'FooOperation',
+            [$inputParam => $inputValue]
+        );
+
+        $list = new HandlerList();
+        $list->setHandler(function ($command) {
+            return "success";
+        });
+        $list->prependBuild(Middleware::tap(function (Command $command) use ($inputParam, $expected) {
+            $this->assertSame($expected, $command[$inputParam]);
+        }));
+        $list->appendValidate(QueryCompatibleInputMiddleware::wrap($service));
+
+        $handler = $list->resolve();
+
+        $result = @$handler($command, new Request('POST', 'https://foo.com'));
+        self::assertSame($result, "success");
+    }
+
+    /**
+     * Data provider for providing top-level command arguments
+     *
+     * @return array
+     */
+    public function getInputs()
+    {
+        return [
+            ['IntParam', '10', 10, 'integer'],
+            ['LongParam', '1000000000000000000000000', 1.0E+24, 'long'],
+            ['FloatParam', '10.0', 10.0, 'float'],
+            ['StringParam', 1234, '1234', 'string'],
+            ['StringParam', 15.5, '15.5', 'string'],
+            ['StringParam', 1000000000000000000000000, '1.0E+24', 'string']
+        ];
+    }
+
+    public function testCastsNestedValues()
+    {
+        $input = [
+            'MapParam' => [
+                'Key1' => [
+                    'DataType' => 123,
+                    'StringValue' => 456
+                ]
+            ],
+            'StructureParam' => [
+                'NestedParam1' => true,
+                'NestedParam2' => '98765'
+            ],
+            'ListParam' => [
+                [
+                    'NestedParam1' => '10',
+                    'NestedParam2' => 20
+                ]
+            ]
+        ];
+
+        $service = $this->generateTestService();
+        $client = $this->generateTestClient($service);
+        $command = $client->getCommand(
+            'FooOperation',
+            $input
+        );
+
+        $list = new HandlerList();
+        $list->setHandler(function ($command) {
+            return "success";
+        });
+        $list->prependBuild(Middleware::tap(function (Command $command) {
```

Comment:
Can this be added to the client's `HandlerList` instead of using a fully rebuilt one?